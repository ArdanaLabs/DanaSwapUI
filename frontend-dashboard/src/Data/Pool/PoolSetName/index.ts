import * as Show_ from "fp-ts/Show"
import * as string from "fp-ts/string"
import * as IO from "io-ts"
import { fromNewtype } from "io-ts-types"
import * as Newtype from "newtype-ts"

export default interface PoolSetName
  extends Newtype.Newtype<{ readonly PoolSetName: unique symbol }, string> {}

export type Type = PoolSetName

export const codec: IO.Type<PoolSetName, string> = fromNewtype<PoolSetName>(
  IO.string
)

export const iso = Newtype.iso<PoolSetName>()

export const Eq = Newtype.getEq<PoolSetName>(string.Eq)

export const Ord = Newtype.getOrd<PoolSetName>(string.Ord)

export const Show: Show_.Show<PoolSetName> = {
  show: (l) => `PoolSetName(${iso.unwrap(l)})`,
}
