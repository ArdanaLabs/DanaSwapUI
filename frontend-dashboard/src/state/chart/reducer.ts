import {
  ActionReducerMapBuilder,
  PayloadAction,
  createReducer,
} from "@reduxjs/toolkit"

import * as E from "fp-ts/Either"
import { RemoteData, fromEither, pending } from "fp-ts-remote-data"

import {
  receivedAggVolume,
  receivedAggLiquidity,
  receivedPoolFees,
  receivedPoolVolume,
  receivedPoolLiquidity,
  receivedPoolTxCount,
  receivedPoolAPY,
  receivedPoolTransactions,
} from "./actions"

import {
  APYChart,
  FeeVolumeChart,
  LiquidityChart,
  TxCountChart,
  VolumeChart,
} from "Data/Chart"
import { FetchDecodeError, FetchDecodeResult } from "Data/FetchDecode"
import * as Transaction from "Data/Transaction"

export type State = {
  aggLiquidity: RemoteData<FetchDecodeError, LiquidityChart.Type>
  aggVolume: RemoteData<FetchDecodeError, VolumeChart.Type>
  poolFees: RemoteData<FetchDecodeError, FeeVolumeChart.Type>
  poolVolume: RemoteData<FetchDecodeError, VolumeChart.Type>
  poolLiquidity: RemoteData<FetchDecodeError, LiquidityChart.Type>
  poolTxCount: RemoteData<FetchDecodeError, TxCountChart.Type>
  poolAPY: RemoteData<FetchDecodeError, APYChart.Type>
  poolTransactions: RemoteData<FetchDecodeError, Transaction.WithTotalValue[]>
}

export const initialState: State = {
  aggLiquidity: pending,
  aggVolume: pending,
  poolFees: pending,
  poolVolume: pending,
  poolLiquidity: pending,
  poolTxCount: pending,
  poolAPY: pending,
  poolTransactions: pending,
}

// side-effectful: function that logs on Left, and returns the typical fromEither
// TODO: move this to chain with the decoder stuff
function fromEitherLogFail<A>({
  payload,
}: PayloadAction<FetchDecodeResult<A>>): RemoteData<FetchDecodeError, A> {
  if (E.isLeft(payload)) {
    console.error("Fetch error:", payload.left)
  }
  return fromEither(payload)
}

export default createReducer(
  initialState,
  (builder: ActionReducerMapBuilder<State>) =>
    builder
      .addCase(receivedAggLiquidity, (state, action) => ({
        ...state,
        aggLiquidity: fromEitherLogFail(action),
      }))
      .addCase(receivedAggVolume, (state, action) => ({
        ...state,
        aggVolume: fromEitherLogFail(action),
      }))
      .addCase(receivedPoolFees, (state, action) => ({
        ...state,
        poolFees: fromEitherLogFail(action),
      }))
      .addCase(receivedPoolVolume, (state, action) => ({
        ...state,
        poolVolume: fromEitherLogFail(action),
      }))
      .addCase(receivedPoolLiquidity, (state, action) => ({
        ...state,
        poolLiquidity: fromEitherLogFail(action),
      }))
      .addCase(receivedPoolTxCount, (state, action) => ({
        ...state,
        poolTxCount: fromEitherLogFail(action),
      }))
      .addCase(receivedPoolAPY, (state, action) => ({
        ...state,
        poolAPY: fromEitherLogFail(action),
      }))
      .addCase(receivedPoolTransactions, (state, action) => ({
        ...state,
        poolTransactions: fromEitherLogFail(action),
      }))
)
