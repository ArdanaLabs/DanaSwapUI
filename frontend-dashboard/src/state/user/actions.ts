import { createAction } from "@reduxjs/toolkit"

import * as O from "fp-ts/Option"

import { SupportedW3ColorScheme, Theme } from "Data/User/Theme"

export const updateUserTheme = createAction<Theme>("user/updateUserTheme")

export const updatePrefersColorScheme = createAction<
  O.Option<SupportedW3ColorScheme>
>("user/updatePrefersColorScheme")
