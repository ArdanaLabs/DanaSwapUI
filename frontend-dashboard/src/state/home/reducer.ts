import {
  ActionReducerMapBuilder,
  PayloadAction,
  createReducer,
} from "@reduxjs/toolkit"

import { RemoteData, fromEither, pending } from "fp-ts-remote-data"

import { CombinedStats } from "Data/Stats/CombinedStats"
import { FetchDecodeError, FetchDecodeResult } from "Data/FetchDecode"

import { receivedCombinedStats } from "./actions"

export type State = RemoteData<FetchDecodeError, CombinedStats>

export const initialState = pending

export default createReducer(
  initialState,
  (builder: ActionReducerMapBuilder<State>) =>
    builder.addCase(
      receivedCombinedStats,
      (
        _state: State,
        action: PayloadAction<FetchDecodeResult<CombinedStats>>
      ) => fromEither(action.payload)
    )
)
