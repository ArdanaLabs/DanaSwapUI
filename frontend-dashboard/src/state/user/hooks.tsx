// import { useCallback } from 'react';
import { useDispatch, useSelector } from "react-redux"

import * as O from "fp-ts/Option"
import { pipe } from "fp-ts/function"

import * as Theme from "Data/User/Theme"

import { AppDispatch, AppState } from "state"
import { UserState } from "./reducer"
import { updateUserTheme } from "./actions"

export function getTheme(
  userState: Pick<UserState, "theme" | "prefersColorScheme">
): Theme.Theme {
  return pipe(
    userState.theme,
    O.alt(
      (): O.Option<Theme.Theme> =>
        O.chain(
          (s: Theme.SupportedW3ColorScheme): O.Option<Theme.Theme> =>
            Theme.fromW3ColorScheme(s)
        )(userState.prefersColorScheme)
    ),
    O.getOrElse((): Theme.Theme => Theme.defaultTheme)
  )
}

export function useUserTheme(): Theme.Theme {
  const user: UserState = useSelector<AppState, UserState>(
    (state) => state.user
  )
  return getTheme(user)
}

export function useUserThemeManager(): [
  Theme.Theme,
  (newTheme: Theme.Theme) => void
] {
  const dispatch = useDispatch<AppDispatch>()
  const theme: Theme.Theme = useUserTheme()

  // const setDarkMode = useCallback(
  //   (darkMode: boolean) => {
  //     dispatch(updateUserDarkMode({ userDarkMode: darkMode }));
  //   },
  //   [dispatch]
  // );
  const setTheme = (newTheme: Theme.Theme): void => {
    dispatch(updateUserTheme(newTheme))
  }

  return [theme, setTheme]
}
