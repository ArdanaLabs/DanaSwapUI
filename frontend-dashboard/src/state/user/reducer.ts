import { ActionReducerMapBuilder, createReducer } from "@reduxjs/toolkit"

import * as O from "fp-ts/Option"

import { SupportedW3ColorScheme, Theme } from "Data/User/Theme"

import { updatePrefersColorScheme, updateUserTheme } from "./actions"

function currentTimestamp(): number {
  return new Date().getTime()
}

export const THEME_LOCALSTORAGE_KEY = "user/theme"

export interface UserState {
  lastUpdateVersionTimestamp?: number
  theme: O.Option<Theme> // current theme
  prefersColorScheme: O.Option<SupportedW3ColorScheme>
  timestamp: number
}

export const initialState: UserState = {
  theme: O.none,
  prefersColorScheme: O.none,
  timestamp: currentTimestamp(), // side-effect
}

export default createReducer(
  initialState,
  (builder: ActionReducerMapBuilder<UserState>) =>
    builder
      .addCase(updateUserTheme, (state, action) => {
        localStorage.setItem(
          THEME_LOCALSTORAGE_KEY,
          action.payload.toString(10)
        )
        const timestamp = currentTimestamp()
        return { ...state, theme: O.some(action.payload), timestamp }
      })
      .addCase(updatePrefersColorScheme, (state, action) => {
        const timestamp = currentTimestamp()
        return { ...state, prefersColorScheme: action.payload, timestamp }
      })
)
