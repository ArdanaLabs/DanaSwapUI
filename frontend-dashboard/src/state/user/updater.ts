import { useEffect } from "react"
import { useDispatch } from "react-redux"

import * as O from "fp-ts/Option"
import { pipe } from "fp-ts/function"

import { Theme, SupportedW3ColorScheme } from "Data/User/Theme"

import { AppDispatch } from "state"
import { updatePrefersColorScheme, updateUserTheme } from "./actions"
import { THEME_LOCALSTORAGE_KEY } from "./reducer"

export default function Updater(): null {
  const dispatch = useDispatch<AppDispatch>()

  useEffect(() => {
    const storedTheme: O.Option<Theme> = pipe(
      localStorage.getItem(THEME_LOCALSTORAGE_KEY),
      O.fromNullable,
      O.chain((s: string): O.Option<Theme> => {
        const i: number = parseInt(s, 10)
        if (Number.isNaN(i) || !(i in Theme)) {
          return O.none
        } else {
          return O.some(i)
        }
      })
    )

    if (O.isSome(storedTheme)) {
      dispatch(updateUserTheme(storedTheme.value))
    }

    // This matching is done in this manner because the W3 spec clearly states
    // that we should expect the current set scheme support to be extended.
    // However, there’s not an easy way to just ‘get’ the string value, unlike
    // `devicePixelRatio`, so getting this match isn’t ‘easy’.

    const mkMatcher = (scheme: SupportedW3ColorScheme): MediaQueryList =>
      window.matchMedia(`screen and (prefers-color-scheme: ${scheme})`)

    const lightMatch: MediaQueryList = mkMatcher(SupportedW3ColorScheme.light)
    const darkMatch: MediaQueryList = mkMatcher(SupportedW3ColorScheme.dark)

    function matchHandler(_evt?: MediaQueryListEvent) {
      let scheme: O.Option<SupportedW3ColorScheme> = O.none
      if (lightMatch.matches) {
        scheme = O.some(SupportedW3ColorScheme.light)
      } else if (darkMatch.matches) {
        scheme = O.some(SupportedW3ColorScheme.dark)
      }
      dispatch(updatePrefersColorScheme(scheme))
    }

    matchHandler()

    lightMatch.addEventListener("change", matchHandler)
    darkMatch.addEventListener("change", matchHandler)

    return () => {
      lightMatch.removeEventListener("change", matchHandler)
      darkMatch.removeEventListener("change", matchHandler)
    }
  }, [dispatch])

  return null
}
