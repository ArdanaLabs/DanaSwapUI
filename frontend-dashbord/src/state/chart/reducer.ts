import { createReducer } from "@reduxjs/toolkit";

import { updateAggVol } from "./actions";

export const initialState: any = {
  aggVolsPerFiveMins: [],
};

export default createReducer(initialState, (builder) =>
  builder.addCase(updateAggVol, (state, action) => {
  })
);
