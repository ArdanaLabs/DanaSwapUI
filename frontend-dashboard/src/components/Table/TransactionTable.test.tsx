import Adapter from "@wojtekmaj/enzyme-adapter-react-17"
import * as Enzyme from "enzyme"
import { shallow } from "enzyme"
import { BrowserRouter } from "react-router-dom"

import * as O from "fp-ts/Option"

import * as PoolSetName from "Data/Pool/PoolSetName"

import { UserState } from "state/user/reducer"
import { TransactionTable } from "."

Enzyme.configure({ adapter: new Adapter() })

const mockUserState: UserState = {
  theme: O.none,
  prefersColorScheme: O.none,
  timestamp: 0,
}

jest.mock("react-redux", () => ({
  useDispatch: () => {},
  useSelector: (): UserState => mockUserState,
}))

let wrapper: Enzyme.ShallowWrapper

describe("Components / Table / TransactionTable", () => {
  beforeEach(() => {
    wrapper = shallow(
      <BrowserRouter basename="/">
        <TransactionTable poolSet={PoolSetName.iso.wrap("foo")} />
      </BrowserRouter>
    )
  })

  it("renders", () => {
    expect(wrapper).toMatchSnapshot()
  })
})
