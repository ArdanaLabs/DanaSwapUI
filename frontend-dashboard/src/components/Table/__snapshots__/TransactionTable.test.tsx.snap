// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components / Table / TransactionTable renders 1`] = `ShallowWrapper {}`;
